src/reclaim.rs

```diff
@@ -0,0 +1,186 @@
+pub(crate) use seize::{Collector, Guard, Linked};
+
+use std::marker::PhantomData;
+use std::ops::Deref;
+use std::sync::atomic::Ordering;
+use std::{fmt, ptr};
+
+pub(crate) struct Atomic<T>(seize::AtomicPtr<T>);
+
+impl<T> Atomic<T> {
+    pub(crate) fn null() -> Self {
+        Self(seize::AtomicPtr::default())
+    }
+
+    pub(crate) fn load<'g>(&self, _: Ordering, guard: &'g Guard<'_>) -> Shared<'g, T> {
+        guard.protect(&self.0).into()
+    }
+
+    pub(crate) fn store(&self, new: Shared<'_, T>, ordering: Ordering) {
+        self.0.store(new.ptr, ordering);
+    }
+
+    pub(crate) unsafe fn into_box(self) -> Box<Linked<T>> {
+        Box::from_raw(self.0.into_inner())
+    }
+
+    pub(crate) fn swap<'g>(
+        &self,
+        new: Shared<'_, T>,
+        ord: Ordering,
+        _: &'g Guard<'_>,
+    ) -> Shared<'g, T> {
+        self.0.swap(new.ptr, ord).into()
+    }
+
+    pub(crate) fn compare_exchange<'g>(
+        &self,
+        current: Shared<'_, T>,
+        new: Shared<'g, T>,
+        success: Ordering,
+        failure: Ordering,
+        _: &'g Guard<'_>,
+    ) -> Result<Shared<'g, T>, CompareExchangeError<'g, T>> {
+        match self
+            .0
+            .compare_exchange(current.ptr, new.ptr, success, failure)
+        {
+            Ok(ptr) => Ok(ptr.into()),
+            Err(current) => Err(CompareExchangeError {
+                current: current.into(),
+                new,
+            }),
+        }
+    }
+}
+
+impl<T> From<Shared<'_, T>> for Atomic<T> {
+    fn from(shared: Shared<'_, T>) -> Self {
+        Atomic(shared.ptr.into())
+    }
+}
+
+impl<T> Clone for Atomic<T> {
+    fn clone(&self) -> Self {
+        Atomic(self.0.load(Ordering::Relaxed).into())
+    }
+}
+
+impl<T> fmt::Debug for Shared<'_, T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:p}", self.ptr)
+    }
+}
+
+impl<T> fmt::Debug for Atomic<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{:p}", self.0.load(Ordering::SeqCst))
+    }
+}
+
+pub(crate) struct CompareExchangeError<'g, T> {
+    pub(crate) current: Shared<'g, T>,
+    pub(crate) new: Shared<'g, T>,
+}
+
+pub(crate) struct Shared<'g, T> {
+    ptr: *mut Linked<T>,
+    _g: PhantomData<&'g ()>,
+}
+
+impl<'g, T> Shared<'g, T> {
+    pub(crate) fn null() -> Self {
+        Shared::from(ptr::null_mut())
+    }
+
+    pub(crate) fn boxed(value: T, collector: &Collector) -> Self {
+        Shared::from(collector.link_boxed(value))
```

Comment:
The fact that a ref to the `Collector` is needed to construct a `Shared` has me a little concerned — does _every_ seize-protected value hold a pointer to the collector? If so, that'll pretty significantly increase the size of a map where the key and value types are small. Say they're `u32`, but now we're sticking in a full `usize` pointer twice in every entry as well!

Reply:
The map holds pointer to the `Linked<T>`, not the value directly, so it shouldn't increase the size?

Reply:
It still increases the overall heap memory use of the map significantly. Previously we'd store (for each item):
```
sizeof(usize) + sizeof(K) + sizeof(usize) + sizeof(V)
```
whereas if each `Linked` _also_ holds a pointer to the container, and every `K` and `V` is a `Linked<_>`, then we're now storing
```
2 * sizeof(usize) + sizeof(K) + 2 * sizeof(usize) + sizeof(V)
```
which is a fairly significant increase in heap memory for a map when `K` and `V` are small, and there are many elements, such as maps to or from integers (which are _very_ common in practice).

Reply:
`Linked` actually stores a lot more than that. `Linked<()>` is 32 bytes, and packs a ton of metadata in there (2 unions). It stores a link to the head of the batch, the recorded epoch value when it was allocated, a link to the next node in the reservation list, the reference count and minimum recorded epoch of the batch, a link to the next node in the batch, and the reservation function :sweat_smile: All of this is pretty fundamental to the reclamation algorithm. Some of the data is inlined from what would instead be a thread-local list in crossbeam-epoch/hazard pointers. Do you think this is a big problem?

Reply:
Oooooof, that's pretty painful. For _every_ value? That's definitely worrying. Consider a map with, say, 100k entries, all of which are `usize`s. Its heap memory use would increase three-fold (from 100k*(2*2*8)=3.2MB to 100k*(2*(8+32+8))=9.6MB). Now, whether it's a deal-breaker, that's harder to say... In its current state, flurry isn't going to be used since its performance is so poor, and at least with this its _performance_ is pretty good, even if its memory footprint is kind of lousy. That said, it would be in Java too, so maybe that's just what we signed up for :p

I think let's still go ahead with this, but maybe add a note about performance and memory characteristics to the docs, and highlight the expected memory use of the map given `N`, `sizeof(K)`, and `sizeof(V)`.

Reply:
Hmmm.. it might be possible to initially only store the epoch value, and then only when a node is _retired_ allocate the 32 bytes needed to keep track of it's reservation status. That does mean we incur an extra pointer indirection anytime reference counts need to be incremented/decremented.

Do you think that would be worthwhile?

Reply:
Does that mean an extra pointer indirection on every read? If so, I'd be curious to see the performance impact. But could very well be worth it, yeah!

Reply:
No not on every read, just in retire and Guard::drop.

Reply:
That definitely sounds promising then!